### DISCUSSION TITLE
query/transform API?

### DISCUSSION THREAD

**Discussion by lygaret** [htdebeer/paru#71] [Ideas]
Hey!

I'm working on using `Paru` as the backbone for a markdown-based database indexer, for example extracting links, task items, tags, etc. and the ability to utilize `paru`s AST for searching and transformations is really great.

One thing I'm curious about - right now, the `Filter` API is very much geared towards I/O via the `pandoc` command, but it would be pretty useful to be able to separate out the concepts of "transforms" from file I/O. I'm doing a good amount of re-parsing to get multiple passes, and that feels unnecessary.

I'm thinking something like:

```ruby
require 'paru/transformer'

input = File.read("./data/index.md")
document = Paru::PandocFilter::Document.from_markdown(input)

# a transformer is stateful, like a filter, but doesn't have output
transformer = Paru::Transformer(document)

# be able to run aribtrary transforms over the document (or any other `has_children?` node)
# each transform block is atomic (semantically, not threads), and transforms the same view of the document, same as `Filter`
transformer.transform do
  with("BulletList > Plain") do |plain|
    plain.parent.replace(plain, TaskListItemNode.from(plain)
  end
end

# transforms stick, so applying a second transform applies to the transformed document from the last one
transformer.transform do
  with("Link") do |link|
    if link.href.start_with? "wiki:"
      link.parent.replace(link, WikiLinkNode.from(link))
    end
  end
end

# it's an imperative reduction of transforms over the document
# after transforms, without rendering and reparsing, we can use the same mechanism to do queries against the ast

tasks = transformer.query("TaskListItem")
db.update_tasks(tasks)

links = transformer.query("WikiLink")
db.update_outgoing_links(links)

# once we're done, we can abandon the changes
# render things out

$stdout << json2pandoc document.to_ast
```

In particular, I'm _not_ suggesting that the `Filter` API is bad; far from it! It's a really, really nice API for the tool it's designed for -- easily slotting into `--filter` calls to pandoc -- it just gets in the way some in trying to use the AST without it.

I think this would probably require:

* **refactoring `Filter`**
   The loop in `Filter::filter` is the core of what a `Transformer` would really need; breaking that loop out into a separate class that `Filter` inherits from and then adds I/O on top of would get pretty far.

   I imagine that `before`/`after`/`stop!` handling stays with the `Filter`, as they only make sense in a single-pass filter anyway, and are trivially replicated by multiple `#transform` calls.

* **opmitized query**
   Right now, doing queries over a document is slow, as it requires doing a search of the entire navigable AST for every query. Once transforms are broken out, they can easily hold type-caches, and other sorts of query-optimizations, that would be no-slower in the one-off case, but could make repeat queries much faster.

* **custom ast nodes**
   This one is more my own thoughts, but the API you've put together for node classes makes it really simple to create custom nodes that represent concepts in the AST. However, `selector` requires that a node name be well known, and there's not documentation around expectations re: `ast_type` and `type`.

   I've been doing this, which is working mostly, but I haven't really stressed it yet:

   ```ruby
   class TaskListItemNode < Plain     
    def type = "TaskNode"  # for matching
    def ast_type = "Plain" # for serialization
   end
   ```

---

Anyhow, hope you're well, and thank you for the cool library!

**Comment by lygaret** [comment]
I meant to add that I'm more that happy to take a pass at this, but I didn't want to get ahead of myself without your blessing @htdebeer :D

**Reply by htdebeer** [reply]
Interesting idea, your transformer! Go ahead and build it, if you want. As long as the code isn't worse than mine and the whole thing is backwards compatible, I'm game. However, I recommend to explore some alternatives designs first, and maybe work out some user scenario's.

You're right that one of the problems with pandoc filters is indeed that they're run in sequence. On the one hand, that makes sense, because you can compose them pandoc-invocation-time and so easily re-use them. However, if you've got a lot of filters, or filters that are expensive to run, combining them into one could speed things up considerably. Thinking ahead, fusing filters—analogous to fusion in functional programming—would improve performance even more. But that might be taking it a step too far :-)

Anyway, in your example, you suggest a `Transformer` class that seems both data structure and behavior? To me, it'd make more sense to think of transformers as a function from `Document` to `Document`. Then chain them with a nice operator, something like:

```ruby
task_list_maker = Transform.new do ... end
wiki_linker = Transform.new do ... end

out_doc = wiki_linker << task_list_maker << in_document
```

Similarly, a `query` method on `Document` seems more natural than one on `Transformer`.

By the way, paru/filter is a very simple wrapper around [pandoc's AST definition](https://hackage.haskell.org/package/pandoc-types-1.22.2/docs/Text-Pandoc-Definition.html). That's also why the selector only knows about these pandoc types, the [Block types](https://hackage.haskell.org/package/pandoc-types-1.22.2/docs/src/Text.Pandoc.Definition.html#Block) and [Inline types](https://hackage.haskell.org/package/pandoc-types-1.22.2/docs/src/Text.Pandoc.Definition.html#Inline).

If you want to create your own types, you need to supply a mapping from your new type to a pandoc AST subtree in terms of those pandoc types.

Finally, if run-time performance is important to your application, ruby and paru might not be the best choice because there's quite some overhead running a filter. That's fine if you're running pandoc with your filters every now and then, but if you want to run them multiple times a second or so, things might get hairy.

**Reply by lygaret** [reply]
Oh, yeah, understood re: performance; it's less about raw wall time than it is the fact that I know it's doing full scans for each transforms :D

Per your composable transform, that's what I would prefer anyway, so I'm going to take a pass. I'll have a super rough draft (enough to burn in effigy) shortly, and we can discuss approaches with something to look at.

Re: custom nodes, that snippet of `TaskItemNode` works as long as the error re: selector types is commented out; the `to_ast` is based on `Plain`, and serializes the right type because we can split `type` and `ast_type` (good call btw).

**Reply by lygaret** [reply]
Also, I wonder if `#query` makes more sense on `Selector`:

```ruby
selector = Selector.new("BulletList > Plain")
results = selector.scan document 
# scan on reified query, rather than query on document
```

that would let the `Document` continue to just represent the AST, move query to something independent, and then eventually one could thread a search context through for optimizations.

**Reply by htdebeer** [reply]
I was thinking analogous to `document.querySelector()` method in the JavaScript HTML DOM, but maybe the `query` method should go on `Node` instead? Then you can query any AST sub tree, everywhere.

I like "document.select(selector)" better than "selector.scan(document)", but that might be my experience with the DOM talking. On the other hand, in OOP languages it seems common to use "container.select(selector)", rather than "selector.scan(container)". The latter form seems more common in imperative languages or utility classes.

**Comment by htdebeer** [comment]
By the way, I've moved paru development to Docker to fix ruby and pandoc version while developing. I didn't write a developer's guide yet, but I have a very short [get started summary](../blob/master/documentation/development.md). 

If you've any questions about this or other things, let me know.
